src/framework/MOM_diag_buffers.F90

```diff
@@ -0,0 +1,472 @@
+end type buffer_3d
+
+!> The base class for the diagnostic buffers in this module
+type, abstract :: diag_buffer_base ; private
+  integer :: is !< The start slot of the array i-direction
+  integer :: js !< The start slot of the array j-direction
+  integer :: ie !< The end slot of the array i-direction
+  integer :: je !< The end slot of the array j-direction
+  real :: fill_value = 0. !< Set the fill value to use when growing the buffer [arbitrary]
+
+  integer, allocatable, dimension(:) :: ids  !< List of diagnostic ids whose slot corresponds to the row in the buffer
+  integer :: length = 0 !< The number of slots in the buffer
+
+  contains
+
+  procedure(a_grow), deferred :: grow !< Increase the size of the buffer
+  procedure, public :: set_fill_value !< Set the fill value to use when growing the buffer
+  procedure, public :: check_capacity_by_id !< Check the size size of the buffer and increase if necessary
+  procedure, public :: set_horizontal_extents !< Define the horizontal extents of the arrays
+  procedure, public :: mark_available !< Mark that a slot in the buffer can be reused
+  procedure, public :: grow_ids !< Increase the size of the vector storing diagnostic ids
+  procedure, public :: find_buffer_slot !< Find the slot corresponding to a specific diagnostic id
+end type diag_buffer_base
+
+!> Dynamically growing buffer for 2D arrays.
+type, extends(diag_buffer_base), public :: diag_buffer_2d; private
+  type(buffer_2d), public, dimension(:), allocatable :: buffer !< The actual 2D buffer which will dynamically grow
+
+  contains
+
+  procedure, public :: grow => grow_2d !< Increase the size of the buffer
+  procedure, public :: store => store_2d !< Store a field in the buffer, increasing as necessary
+end type diag_buffer_2d
+
+!> Dynamically growing buffer for 3D arrays.
+type, extends(diag_buffer_base), public :: diag_buffer_3d ; private
+  type(buffer_3d), public, dimension(:), allocatable :: buffer !< The actual 2D buffer which will dynamically grow
+  integer :: ks !< The start slot in the k-dimension
+  integer :: ke !< The last slot in the k-dimension
+
+  contains
+
+  procedure, public :: set_vertical_extent !< Set the vertical extents of the buffer
+  procedure, public :: grow => grow_3d !< Increase the size of the buffer
+  procedure, public :: store => store_3d !< Store a field in the buffer, increasing as necessary
+end type diag_buffer_3d
+
+contains
+
+!> Signature for the grow methods on n-dimension diagnostic buffer types
+subroutine a_grow(this)
+  class(diag_buffer_base), intent(inout) :: this !< The diagnostic buffer
+end subroutine
+
+!> Set the fill value to use when growing the buffer
+subroutine set_fill_value(this, fill_value)
+  class(diag_buffer_base), intent(inout) :: this !< The diagnostic buffer
+  real,                    intent(in)    :: fill_value !< The fill value to use when growing the buffer [arbitrary]
```

Comment:
I believe that this subroutine is missing the line `this%fill_value = fill_value`, without which it does not do anything.

Reply:
Thanks for pointing this out (line was lost in a refactor). I added a test to make sure that this actually does what we expect